test/Interop/Cxx/templates/function-template.swift

```diff
@@ -39,6 +39,97 @@ FunctionTemplateTestSuite.test("constLvalueReferenceToBool<T> where T == Bool")
   expectFalse(constLvalueReferenceToBool(false))
 }
 
+FunctionTemplateTestSuite.test("callFunction<T> where T == () -> ()") {
+  var value = false
+  callFunction({() in value = true})
+  expectTrue(value)
+}
+
+FunctionTemplateTestSuite.test("callFunction<T> where T == () -> Bool") {
+  var value = false
+  callFunction({() in value = true; return true})
+  expectTrue(value)
+}
+
+FunctionTemplateTestSuite.test("callFunctionWithParam<T, A> where T == (Int) -> (), A == Int") {
+  var value = 42
+  callFunctionWithParam({(x: Int) in value = x}, 24)
+  expectEqual(value, 24)
+}
+
+FunctionTemplateTestSuite.test("callFunctionWithParam<T, A> where T == (Int) -> Bool, A == Int") {
+  var value = 42
+  callFunctionWithParam({(x: Int) in value = x; return true}, 24)
+  expectEqual(value, 24)
+}
+
+FunctionTemplateTestSuite.test("callFunctionWithParam<T, A> where T == (CxxClass) -> (), A == CxxClass") {
+  var value: Int32 = 42
+
+  // Read x via direct field access
+  callFunctionWithParam({(c: CxxClass) in value = c.x}, CxxClass(x: 24))
+  expectEqual(value, 24)
+
+  // Read x via getter method
+  callFunctionWithParam({(c: CxxClass) in value = c.getX()}, CxxClass(x: 22))
+  expectEqual(value, 22)
+}
+
+// TODO: this likely needs more design work
+// FunctionTemplateTestSuite.test("callFunctionWithParam<T, A> where T == (()) -> (), A == ()") {
+//   var value = 42
+//   // None of these are quite right:
+//   // callFunctionWithParam({(_: ()) in value = 24}, ())
+//   // callFunctionWithParam({(_: Void) in value = 24}, ())
+//   // callFunctionWithParam({() in value = 24}, ())
+//   expectEqual(value, 24)
+// }
+
+FunctionTemplateTestSuite.test("callFunctionWithReturn<T, B> where T == () -> Int, B == Int") {
+  let val1: Int = callFunctionWithReturn({() in 111})
+  expectEqual(val1, 111)
+  let val2: Int = callFunctionWithReturn({() -> Int in 222})
+  expectEqual(val2, 222)
+  // Cannot infer return type without val3 explicitly annotated as Int
+  // let val3 = callFunctionWithReturn({() -> Int in 333})
+  // expectEqual(val3, 333)
+
+  // This works:
+  expectEqual(callFunctionWithReturn({() in 444}), 444)
+  // This too:
+  expectEqual(callFunctionWithReturn({() in true}), true)
+  // But this does not (probably a quirk of expectTrue())
+  // expectTrue(callFunctionWithReturn({() in true}))
+}
+
+FunctionTemplateTestSuite.test("callFunctionWithReturn<T, B> where T == () -> CxxClass, B == CxxClass") {
+  let c: CxxClass = callFunctionWithReturn({() in CxxClass(x: 42)})
+  expectEqual(c.x, 42)
+
+  let d: CxxClass = callFunctionWithReturn({() in c})
+  expectEqual(d.x, 42)
+}
+
+FunctionTemplateTestSuite.test("indirectlyCallFunction<T> where T == () -> ()") {
+  var value = 42
+
+  indirectlyCallFunction({() in value = 111});
+  expectEqual(value, 111)
+
+  indirectlyCallFunctionTemplate({() in value = 222});
+  expectEqual(value, 222)
+}
+
+FunctionTemplateTestSuite.test("indirectlyCallFunction<T> where T == () -> ()") {
```

Comment:
I am not sure how blocks are handled but is it possible to import a block from Obj-C? If so, is that imported as a closure? Could we instantiate a C++ template with a block that we imported from Obj-C?

Reply:
I _think_ that should be possible. I'll add a test for that.

Reply:
Something that _doesn't_ work yet, though, are C function pointers, because pointers in general are still quite broken.

Reply:
So it turns out C function pointers work; I've updated the test suite to cover those too. It's only _optional_ (function) pointers that are broken.